Logger process trigger

nordic-krch opened the pull request:
Two features are introduced:
- option to wake up thread which is responsible for log processing. Thread can be waken up when number of buffered messages exceeds configurable threshold. Thread id is provided during log initialization and can be null if feature is not used - threshold set to 0.
- added optional internal logger thread. If enabled, logger thread is created, thread is initializing the logger and process messages. If there is no pending messages thread goes to sleep for configurable period. Thread might be waken up using periodically or using threshold. Stack size set to 768 bytes (stack analysis shown that 576 bytes were used on nrf52).

Logger sample updated to showcase new features.

nordic-krch committed:
logging: Logger to wake up logs processing thread

Added configurable threshold of number of buffered log messages
on which log wakes up thread which processes buffered logs. Thread
ID is provided during logger initialization. Feature is optional
and can be disabled by setting CONFIG_LOG_PROCESS_TRIGGER_THR to 0.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add internal thread for log processing

When enabled, logger is creating own thread which processes buffered
logs. When no logs to process, thread sleeps for configurable period.
Thread can be waken up if number of buffered log messages exceeds
configured threshold. Logging sample aligned to use new feature.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
cc: @jukkar

jukkar commented on subsys/logging/log_core.c:
```diff
@@ -416,3 +416,23 @@ u32_t log_filter_get(struct log_backend const *const backend,
+
+	while (1) {
+		if (log_process(false) == false) {
+			k_sleep(CONFIG_LOG_PROCESS_THREAD_SLEEP_MS);
```
I think sleeping here is not good. We should have a way to be notified here that there are messages waiting to be processed, k_poll() could be used here for example.

nordic-krch replied:
There is notifying option: number of buffered messages are tracked and when exceeds the threshold task is waken up. Sleeping is added for case when logs are not arriving continuously and some could stuck for long period. I added threshold to avoid calling k_wakeup() on every log entry because that degrades performance of the logger.

jukkar replied:
Do you mean that this thread is put to sleep and then waken up by some other thread?

Having a k_sleep() here means that basically you are polling the logger. I think this is not really good to do it like this. Better would be that when things exceed some threshold, the logging APIs will trigger a pollable event, which causes this reading thread to be waken up and process the data. After processing this data, the thread starts to wait another events. This is a bit more code than plain sleep, but is better architecturally.

nordic-krch replied:
yes, PR contains 2 commits, first commit (https://github.com/zephyrproject-rtos/zephyr/pull/8704/commits/949505964e4ad2889718724c10de162aa3a5b277) adds ability to wake up a thread when threshold is exceeded. Second commit (adding internal thread) is using this feature but additionally adds sleep period in the thread. Sleeping is added to ensure that buffered logs are processed (and presented to the developer) even when logs stopped arriving. Let say that threshold is set to 20 and you have 19 logs in a buffer and next log will not come for dozens of seconds. You may get confused in that situation if you except to see given log (e.g. you got connected, but log does not print "connected").

Sleep period is configurable (maybe i should add option where sleep period 0 sets forever sleeping).

jukkar commented:
BTW, do we really need to have "subsys:" in the commit subject? I think this is redundant as "logging" is your subsystem so there is no need to add 8 characters more to commit subject, that give no real information.

nashif commented on subsys/logging/log_core.c:
```diff
@@ -393,3 +416,23 @@ u32_t log_filter_get(struct log_backend const *const backend,
+	}
+}
+
+K_THREAD_DEFINE(log_process_thread_id, STACKSIZE, log_process_thread,
```
call this 'log_process_thread', then it appears as such when we list threads by name (using stringify)

nordic-krch commented on subsys/logging/Kconfig:
```diff
@@ -209,11 +209,37 @@ config LOG_INPLACE_PROCESS
+	  because one log operation can be interrupted by another one in higher
+	  priority context.
+
+config LOG_PROCESS_TRIGGER_THR
```
THR was supposed to mean threshold. I've extended it to threshold to avoid confusion and modified prompt to "Amount of buffered logs which triggers processing thread."

pizi-nordic commented on subsys/logging/log_core.c:
```diff
@@ -173,14 +187,22 @@ static u32_t timestamp_get(void)
 }
 
-int log_init(void)
+int log_init(k_tid_t process_tid)
```
We definitely have to separate logger initialization from thread-based processing initialization.
The logger must be be able to log data very early (ideally just after entering C code), while threads could start after system is fully initialized.

Ideally we should end with a call to log_init() in Zephyr startup code and then application might decide how it is going to process gathered data (in the idle, dedicated thread or somewhere else).

nordic-krch replied:
good point, I've extracted setting thread to external function. log_init() is without parameters and initializes fifo and does not touch thread part. I checked that it will play nicely with #8742

nordic-krch committed:
doc: subsys: logging: internal thread and thread wake up

Documenting new logger features: waking up processing thread
and internal logger processing thread.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pizi-nordic commented on doc/subsystems/logging/logger.rst:
```diff
@@ -109,6 +109,13 @@ be processed by printk. Longer strings are trimmed.
+messages reaches the threshold dedicated thread (see :cpp:func:`log_thread_set`)
+is waken up.
+
+:option:`CONFIG_LOG_PROCESS_THREAD`: When enabled, logger is creating own thread
```
Could we add an option which will set priority of internal thread?

nordic-krch replied:
added configurable priority and stack size.

pizi-nordic commented on samples/subsys/logging/logger/src/main.c:
```diff
@@ -218,21 +218,22 @@ void main(void)
-	while (true == LOG_PROCESS()) {
-	}
+	/* Ensuring that log thread will process. */
+	k_sleep(sleep_period);
```
Could we have a log_flush() or something else which will block current thread until log queue is empty?

nordic-krch replied:
I've added log_buffered_cnt() to log_ctrl.h api which returns number of currently pending log messages. It is later on used in the example to pend on log flush

pizi-nordic commented on subsys/logging/Kconfig:
```diff
@@ -209,11 +209,37 @@ config LOG_INPLACE_PROCESS
+	  up. Log processing thread ID is provided during log initialization.
+	  Set 0 to disable the feature.
+
+config LOG_PROCESS_THREAD
```
I suggest 2 different options:

LOG_PROCESS_THREAD: Allow for log processing thread, enables other LOG_PROCESS_THREAD_* options.
LOG_PROCESS_THREAD_INTERNAL: Enables internal thread for log processing.

pizi-nordic commented on subsys/logging/log_core.c:
```diff
@@ -181,6 +196,7 @@ int log_init(void)
 	timestamp_func = timestamp_get;
 	log_output_timestamp_freq_set(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
 
+
```
Is that empty line needed?

nordic-krch replied:
no

pizi-nordic commented on subsys/logging/Kconfig:
```diff
@@ -209,11 +209,37 @@ config LOG_INPLACE_PROCESS
+	  number of buffered messages exceeds the threshold (see
+	  LOG_PROCESS_TRIGGER_THR).
+
+config LOG_PROCESS_THREAD_SLEEP_MS
```
There sould be a note, that this option affects only internal thread.

nordic-krch replied:
yep

dbkinder reviewed (approved):
+1 for doc changes as written.

pizi-nordic commented on doc/subsystems/logging/logger.rst:
```diff
@@ -109,6 +109,13 @@ be processed by printk. Longer strings are trimmed.
 
+:option:`CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD`: When number of buffered log
+messages reaches the threshold dedicated thread (see :cpp:func:`log_thread_set`)
+is waken up.
```
IMHO it might be worth to mention, that this affects internal thread if it is enabled.

pizi-nordic commented on doc/subsystems/logging/logger.rst:
```diff
@@ -221,6 +228,13 @@ be used to change maximal severity level for given module. Module is identified
+which will wake up processing thread when certain amount of log messages are
+buffered (see :option:`CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD`). It is also
+possible to enable internal logger thread (see
+:option:`CONFIG_LOG_PROCESS_THREAD`). In that case logger is initialized and
```
Last sentence is unclear to me. 
What about something like this: "In that case logger thread is initialized and log messages are processed implicitly."?

pizi-nordic commented on samples/subsys/logging/logger/src/main.c:
```diff
@@ -189,15 +194,18 @@ static void external_log_system_showcase(void)
-	(void)err;
+void log_demo_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	k_sleep(100);
```
Why this k_sleep() is needed?

nordic-krch replied:
it can be removed once #8842 goes in. It's about allowing log to initialize since log thread is on low priority and initialization is there.

pizi-nordic replied:
@nashif: Could you please merge #8842?
@nordic-krch: Could you please rebase this after #8842 will be merged?

nordic-krch replied:
done

pizi-nordic commented on subsys/logging/log_core.c:
```diff
@@ -203,6 +219,18 @@ int log_init(void)
 	return 0;
 }
 
+void log_thread_set(k_tid_t process_tid)
```
Should we allow to replace logger internal thread?

nordic-krch replied:
It is exposed in case user does not use internal thread. I can add assert there to check if proc_tid before setting is null.
